### ISSUE TITLE
Tracking Issue for directory handles

### ISSUE THREAD

**Issue by the8472** [rust-lang/rust#120426]
Feature gate: `#![feature(dirfd)]`

This is a tracking issue for directory handles. Such handles provide a stable reference to an underlying filesystem object (typically directories) that are less vulnerable to TOCTOU attacks and similar races. These security properties will be platform-dependent. Platforms that don't provide the necessary primitives will fall back to operations on absolute paths.

Additionally they may also provide performance benefits by avoiding repeated path lookups when performing many operations on a directory.

Sandboxing is a non-goal. If a platform supports upwards path traversal via `..` or symlinks then directory handles will not prevent that. Providing `O_BENEATH`-style traversal is left to 3rd-party crates or future extensions.

### Public API

```rust
impl Dir {
     pub fn open<P: AsRef<Path>>(&self, path: P) -> Result<File>
     /// This could be put on OpenOptions instead
     pub fn open_with<P: AsRef<Path>>(&self, path: P, options: &OpenOptions) -> Result<File>
     pub fn create_dir<P: AsRef<Path>>(&self, path: P) -> Result<()>
     pub fn rename<P: AsRef<Path>, Q: AsRef<Path>>(&self, from: P, to_dir: &Self, to: Q) -> Result<()>
     pub fn remove_file<P: AsRef<Path>>(&self, path: P) -> Result<()>
     pub fn remove_dir<P: AsRef<Path>>(&self, path: P) -> Result<()>
     pub fn symlink<P: AsRef<Path>, Q: AsRef<Path>>(&self, original: P, link: Q)
     
     /// ... more convenience methods
}

impl DirEntry {
    pub fn open(&self) -> Result<File>
    /// This could be put on OpenOptions instead
    pub fn open_with(&self, options: &OpenOptions) -> Result<File>
    pub fn remove_file(&self) -> Result<()>
    pub fn remove_dir(&self) -> Result<()>
}
```

### Steps / History

- [x] ACP: https://github.com/rust-lang/libs-team/issues/259
- [ ] portable, insecure `openat` emulation based on `Path`s
- [ ] real impls for various platforms
- [ ] `getdents` to get free conversion between dirfds and `ReadDir`
- [ ] add more `*at` calls
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by the8472** [comment]
During the ACP a survey of platform support was requested to make sure that fallbacks won't be needed on Tier-1 platforms.


# libc/syscall survey

For restricted conversion between `*DIR` and a file descriptor we need `fdopendir` + `dirfd`. For free conversion we need an alternative to `readdir` that operates on file descriptors.

Presence of `*at` functions was surveyed by looking at the `libc` crate, so this may be incomplete.

## Tier 1

### Windows

* readdir: `GetFileInformationByHandleEx`
* openat: `NtCreateFile`
* renameat: `SetFileInformationByHandle` + `FILE_RENAME_INFORMATION`
* unlinkat: `SetFileInformationByHandle` + `FILE_DISPOSITION_INFO_EX`
* mkdirat: `NtCreateFile`
* symlinkat: ?
* fstatat: ?

### Linux

* readdir: `getdents`
* has fdopendir, dirfd, openat, renameat, unlinkat, mkdirat, symlinkat, fstatat

### macos

* readdir: `getattrlistbulk`
* has fdopendir, dirfd, openat, renameat, unlinkat, mkdirat, symlinkat, fstatat

## Tier 2

### freebsd, illumos, netbsd, solaris

* readdir: `getdents`
* have fdopendir, dirfd, openat, renameat, unlinkat, mkdirat, symlinkat, fstatat

### fuchsia

* readdir: `ReadDirents`
* has fdopendir, dirfd, openat, renameat, unlinkat, mkdirat, symlinkat, fstatat

### ios

* readdir: `getattrlistbulk`
* has fdopendir, dirfd, openat, renameat, unlinkat, mkdirat, symlinkat, fstatat


### wasm-emscripten

* readdir: [getdents](https://github.com/emscripten-core/emscripten/blob/72c3be342861765b620ea0d00e192aa2ac19b744/src/library_syscall.js#L703C3-L703C23)
* has fdopendir, dirfd, openat, renameat, unlinkat, mkdirat, symlinkat, fstatat

### wasm-wasi

* readdir: [fd_readdir](https://github.com/WebAssembly/WASI/blob/main/legacy/preview1/docs.md#-fd_readdirfd-fd-buf-pointeru8-buf_len-size-cookie-dircookie---resultsize-errno)
* has fdopendir, dirfd, openat, renameat, unlinkat, mkdirat, symlinkat, fstatat

### redox

`libc` [only lists](https://docs.rs/libc/0.2.152/x86_64-unknown-redox/libc/index.html) renameat, unlinkat, symlinkat, fstatat

The [redox syscall crate](https://docs.rs/redox_syscall/latest/syscall/call/index.html) lists no *at calls at all? 

### uefi

Listed as no_std in the platform support documentation. Somehow has some pieces in std anyway. No `libc` support either.

## Not checked

- tier 3 targets
- fortanix

**Comment by jackpot51** [comment]
Regarding redox, relibc is the right place to look for support. We should be able to support the same list of functions as Linux and the BSDs.

**Comment by chloekek** [comment]
The ACP mentions “it may be possible to add the Dir methods directly to ReadDir instead”. This is problematic because [`ReadDir: !From<OwnedFd>`](https://github.com/rust-lang/rust/issues/56777). Being able to open a file with `File::open`, then check its file type using `Metadata::file_type` and convert it to a `Dir` object if it is a directory would be very useful IMO, for instance to hash a tree of files.

**Comment by RalfJung** [comment]
From a Miri perspective I'd love to see this, it would let us implement support for emulation of `openat`  and similar functions in a cross-platform way. :)

> portable, insecure openat emulation based on Paths

I am somewhat surprised to see this TODO item. In other areas Rust chose to not provide lower-quality fallback implementations, e.g. famously Rust atomics are always hardware atomics and never lock emulated. I can't find discussion of this point in the [ACP](https://github.com/rust-lang/libs-team/issues/259) either. What is the reasoning here for preferring an insecure emulation over a security guarantee stated in the API?

**Comment by the8472** [comment]
Operating on a directory is a fairly basic operation, like opening a `File`. If we returned an error that basically says "we could do that, just *the way* you're opening it is unsupported on this system" then it would become less portable and likely only interesting for tools like `sudo` which won't support such platforms anyway.

`Dir` isn't just about security. It provides a handle to a directory that should keep working even if the directory gets renamed (granted, fallbacks don't provide this). It helps with performance. It allows replacing global process state (the *current working directory*) with local state. It can prevent non-malicious races.
So applications that aren't security-sensitive would still want to use `Dir` in their filesystem code.

Additionally, the platforms where `openat` isn't available are more likely to be single-user systems without privilege separation, on those fallbacks wouldn't be an issue anyway. Afaik UEFI has none of that.

That said, maybe we can add a `fail_closed` flag, `impl Dir { const IS_RACEFREE: bool }` or something like that for applications that need some extra assurances.


Also, this wouldn't be a runtime fallback but a platform-based fallback. All tier-1 platforms will use real handles and fail if they're not available. E.g. if something blocks `openat` via seccomp on linux I do not intend to try `open`.

**Comment by addisoncrump** [comment]
> Also, this wouldn't be a runtime fallback but a platform-based fallback.

Can this be completed with a compiler warning for these platforms? Since this will be something included in the target type, I would imagine so.

**Comment by the8472** [comment]
@rustbot ping fuchsia

Since it appears that [functions being available in the libc crate is not a reliable indicatior that they actually work](https://github.com/rust-lang/rust/pull/124530), is there a better overview of the available posix APIs? I'm interested in the file descriptor APIs listed in https://github.com/rust-lang/rust/issues/120426#issuecomment-1913167876

**Comment by erickt** [comment]
@the8472 how odd, we do support it at the Fuchsia layer, maybe we're not exposing it through fdio. We'll do some investigation and report back.

**Comment by erickt** [comment]
@the8472 got confirmation from our team that we do support `dirfd` on fuchsia, so this is just a bug we'll get fixed.

**Comment by cgwalters** [comment]
> Sandboxing is a non-goal. If a platform supports upwards path traversal via .. or symlinks then directory handles will not prevent that. Providing O_BENEATH-style traversal is left to 3rd-party crates or future extensions.

I guess that's a reasonable position, but I would also say that for a lot of the code that would be intentionally switching to these APIs, it often makes sense (when rewriting) to just go to such a sandboxed approach anyways.

I'm a bit surprised this tracking issue doesn't reference [cap-std](https://docs.rs/cap-std/latest/cap_std/) as prior art - I'm sure it's *known* but we definitely should think about it intentionally.

Especially because when considering things like platform support, there's a lot of prior art that landed there for some raw functionality.

A very important choice to make here is whether the directory is opened as `O_PATH` (on Linux) or not - cap-std does it this way and I think that's right, but it does have some implications.

**Comment by ChrisDenton** [comment]
> I'm a bit surprised this tracking issue doesn't reference [cap-std](https://docs.rs/cap-std/latest/cap_std/) as prior art

If you click through to the API Change Proposal (ACP) you'll see it's prominently mentioned. The tracking issue is just a summary of the proposed API as currently envisioned.

**Comment by the8472** [comment]
> A very important choice to make here is whether the directory is opened as O_PATH (on Linux) or not - cap-std does it this way and I think that's right, but it does have some implications.

I was thinking of opening them with `O_RDONLY` by default to get a free conversion to `ReadDir` and fall back to `O_PATH` when a directory isn't readable. Though that probably wouldn't be guaranteed anyway since this is rather platform-specific behavior. Some have `O_PATH` or `O_EXEC`, others don't.
What implications did you have in mind?

**Comment by dead-claudia** [comment]
> ## uefi
>
> Listed as no_std in the platform support documentation. Somehow has some pieces in std anyway. No `libc` support either.

#100499 is tracking adding std support for UEFI environments where applicable, with the ultimate goal of making it no longer `#![no_std]` anymore.

**Comment by erickt** [comment]
@the8472: lost track of this, but #136213 should fix fuchsia using `dirfd` and a few other libc filesystem functions.

**Comment by Ayush1325** [comment]
> > ### uefi
> > Listed as no_std in the platform support documentation. Somehow has some pieces in std anyway. No `libc` support either.
> 
> [#100499](https://github.com/rust-lang/rust/issues/100499) is tracking adding std support for UEFI environments where applicable, with the ultimate goal of making it no longer `#![no_std]` anymore.

It is listed with `?`, which means work-in-progress std.

**Comment by RalfJung** [comment]
Looks like Go is getting something similar, though their version does provide directory isolation: https://go.dev/blog/osroot.

**Comment by addisoncrump** [comment]
> Looks like Go is getting something similar

Yes, this was deemed effectively necessary to mitigate certain attacks in the Go space, including those listed in the original issue. A few years ago, I raised some concerns in a security advisory discussion about potential hardening issues present in Rust (and subsequently downstream crates) as a result of not having this supported in std. I would link to this if I could find it, but GitHub does not have good search functionality here... 😅 I'll see if I can't find it.

There are of course alternatives in downstream crates, but it would be good to have support for this internally IMO.

**Comment by RalfJung** [comment]
Solving those race conditions is a separate question from whether things should be isolated with the original root somehow, i.e. whether there should be protection against directory traversal attacks. Sometimes, of course, directory traversal is what one wants, and so far I think the plan for the new APIs is to fix the race conditions, but directory traversal is considered out-of-scope.

**Comment by ChrisDenton** [comment]
> There are of course alternatives in downstream crates, but it would be good to have support for this internally IMO.

Well the ACP has been accepted, it's just waiting on an implementation. Anyone can get started with that if they like. Even if it just implements `open` on one platform then that would be a good enough starting point in a nightly-only API.

**Comment by addisoncrump** [comment]
100%, and the traversal problem would already be handled by this change partially since the `OpenOptions` may include `O_NOFOLLOW` (AFAIK).

**Comment by ebkalderon** [comment]
I stumbled upon an interesting [discussion on StackOverflow](https://stackoverflow.com/a/76183829) regarding [`flock(2)`](https://man7.org/linux/man-pages/man2/flock.2.html) advisory file locks on UNIX-like platforms. Apparently, it is possible to acquire an exclusive lock on a `dirfd`, just like one would on regular files.

If that's true, perhaps we could add `.lock()` and `.try_lock()` methods to the existing `std::fs::Dir` API with analogous behavior to [`std::fs::File::lock()`](https://doc.rust-lang.org/std/fs/struct.File.html#method.lock) and [`std::fs::File::try_lock()`](https://doc.rust-lang.org/std/fs/struct.File.html#method.lock) (#130994)?

I haven't personally confirmed the veracity of this claim, nor have I tested its applicability to other operating systems, e.g. Windows. Just wanted to surface the idea here for discussion! 😄 

---

**EDIT:** I can confirm that opening a directory handle with `File::open()` and then calling `.lock()` and/or `.lock_shared()` does indeed behave exactly as described above, on Linux at least. 👍 I haven't compared the behavior on macOS nor *BSD yet.

**Comment by tmccombs** [comment]
Another use case for this is it would allow opening files with absolute paths that exceed the maximum allowed path without having to change the process's working directory, by using `openat` (or equivalent) along the components of the path.

See https://github.com/BurntSushi/walkdir/issues/23, https://github.com/uutils/coreutils/issues/2949, etc. 

Although that would require the ability to open a subdirectory from the directory handle.

---

I also see that a lot of the API from the ACP isn't in the current API, could I help with implementing some of that?

**Comment by RalfJung** [comment]
The sketched API for `impl Dir` does not have a `metadata` method. Is that expected? I would have expected "get the metadata of that directory" to be a supported operation.

**Comment by RalfJung** [comment]
In Miri we'd like to use directory handles to back up the functions used by `remove_dir_all`. It seems like that requires some operations that are not part of the listed API:
- `fdopendir`: create an `fs::ReadDir` from an `fs::Dir`
- `openat` with `libc::O_NOFOLLOW | libc::O_DIRECTORY`: open an `fs::Dir` relative to an `fs::Dir`, without following symbolic links. ~~(After reading the docs I am surprised that this uses `O_DIRECTORY` rather than `O_PATH`.)~~

Are there plans to add operations like that?

**Comment by the8472** [comment]
> fdopendir: create an fs::ReadDir from an fs::Dir

Yeah the desire for conversion is mentioned in https://github.com/rust-lang/rust/issues/120426#issuecomment-1913167876 this can probably be a `TryFrom` or an inherent method

> openat with libc::O_NOFOLLOW | libc::O_DIRECTORY

Should be achievable via `Dir::open_with` combined with `OpenOptionsExt`

> In Miri we'd like to use directory handles to back up the functions used by remove_dir_all.

cfg(miri) std uses the fallback impl in fs::common though, which doesn't use any of those things.

**Comment by RalfJung** [comment]
> cfg(miri) std uses the fallback impl in fs::common though, which doesn't use any of those things.

We'd like to remove that `cfg(miri)`. It's always bad when Miri executes different code than the compiled program.

**Comment by RalfJung** [comment]
Also seems like it'd be nice if `remove_dir_all` could itself be implemented on top of directory handles rather than having to directly invoke libc. :)

**Comment by RalfJung** [comment]
> Should be achievable via Dir::open_with combined with OpenOptionsExt

So there are no plans for a portable "open without following symlinks"? Well I guess I'll have to ACP that once enough of the rest of the infrastructure is in place. ;)

**Comment by ChrisDenton** [comment]
I would prefer we get a minimal API implemented first instead of trying to do everything at once (and thus getting eternally stalled).

**Comment by RalfJung** [comment]
Fully agreed, I was thinking of what to do once we have that. Maybe it's a bit early for that.

**Comment by RalfJung** [comment]
`Dir` has a `metadata` method that corresponds to `File::metadata`, i.e., "get the metadata of this handle".

However, `Dir` should also have methods corresponding to `fs::metadata`/`fs::symlink_metadata`, i.e., "get the metadata of something relative to this directory".

Any ideas for how we'd best de-conflict those method names?

**Comment by the8472** [comment]
The overly clever option is to remove the self-metadata and instead require the user to pass "." path to get the metadata for the current dir.

Otherwise:

- `metadata`
- `entry_metadata`
- `entry_metadata_nofollow` or just `symlink_metadata`

**Comment by RalfJung** [comment]
In https://github.com/rust-lang/rust/pull/163024 I went for `metadata_at`, but I'm happy to rename.

I considered `entry_*` but since this isn't about `DirEntry` it felt a bit odd.

**Comment by RalfJung** [comment]
Also regarding what you said [here](https://github.com/rust-lang/rust/pull/163007#issuecomment-5742613528)

> Maybe you're thinking of the conversions between `ReadDir` and `Dir`? Those are problematic, need to be fallible and we need to reimplement `ReadDir` to use `getdents` instead of `readdir` because `fdopendir` says
> 
> > Upon successful return from fdopendir(), the file descriptor is under the control of the system, and if any attempt is made to close the file descriptor, or to modify the state of the associated description, other than by means of closedir(), readdir(), readdir_r(), rewinddir(), or [XSI] [Option Start] seekdir(), [Option End] the behavior is undefined.
> 
> I assume that "undefined" really just means "unspecified", but... anyway, different topic.

Closing can easily be prevented by having the conversion consume the `Dir`. And `Dir` itself does not provide any way to modify the associated file description, so -- why do we need to switch to `getdents`? Is it because I could have two `Dir` for the same underlying file description and convert them both into `ReadDir`? Hm... maybe we should not have `try_clone()` or `From<OwnedFd>` then, and have an invariant that a `Dir` is the *only* file descriptor for its underlying file description... (This is an invariant that a while ago I argued `OwnedFd` should maintain, but then we went with a different choice. Maybe that was a mistake...)

**Comment by the8472** [comment]
The Dir<>Fd conversion is worth more imo since it allows interop with rustix, cap-std and other crates.

A ReadDir<>Dir conversion is nice-to-have but we have options

- reimplement `ReadDir` by using getdents which doesn't have precondition (since it's not a userspace struct)
- check with individual libcs what the worst is that can happen
- re-open the fd via `openat(dirfd, ".")` during the conversion, so fallible conversion instead of `From`/`Into`

**Comment by RalfJung** [comment]
> re-open the fd via openat(dirfd, ".") during the conversion, so fallible conversion instead of From/Into

Oh I like that one -- basically creating a new file description that we know is not shared with anything.

So maybe the API for that should be just
```rust
impl Dir {
  pub fn read_dir(&self, path: impl AsRef<Path>) -> io::Result<ReadDir>;
}
```
and via "." that then also covers the conversion part.

**Comment by the8472** [comment]
Sure, that's the baseline, but if we can offer the conversion without syscalls then users can use the types more flexibly/at lower cost.

**Comment by RalfJung** [comment]
So, `Dir::read_dir` is tricky.

What about `ReadDir::dir`?
```rust
impl ReadDir {
    /// Returns a directory handle for the directory that is being iterated.
    pub fn dir(&self) -> io::Result<Dir>;
}
```
On Linux and macOS (and FreeBSD and Solarish and probably a few more), this is trivial to implement by calling `dirfd`+`dup`.

On Windows, this is currently hard to implement; `ReadDir` would have to be rewritten to use a different directory enumeration API.

But is this an API you'd consider having? I am honestly not sure what the use-case would be, other than the fact that it exposed the corresponding POSIX operation `dirfd` -- which is why Miri cares about it. And would it be okay to land an unstable first version that's only race-free on Unixes with `dirfd`? Of course the Windows version should be made race-free before stabilization.

**Comment by the8472** [comment]
> On Linux and macOS (and FreeBSD and Solarish and probably a few more), this is trivial to implement by calling dirfd+dup.

Nope, same issue and same possible solutions as https://github.com/rust-lang/rust/issues/120426#issuecomment-5743929528

> or to modify the state of the associated **description** other than by means [...] the behavior is undefined.

It's not just the descriptor you shouldn't share with arbitrary callers, but the description.


> But is this an API you'd consider having? 

Yes, going between `ReadDir` and `Dir` is useful because it lets the caller do stuff on the directory that's being iterated in a TOCTOU-resilient manner, without having to reimplement the whole `Dir` interface on `ReadDir`

**Comment by RalfJung** [comment]
> Nope, same issue and same possible solutions as https://github.com/rust-lang/rust/issues/120426#issuecomment-5743929528

I don't think so? There's nothing you can do with a `Dir` that would break the `readdir`. We should maybe explicitly document that it's not allowed to pass that FD to `fdopendir`.

But, `ReadDir::dir` could also do `dirfd`+`open` rather than `dirfd`+`dup`. I just think that's not necessary.

**Comment by the8472** [comment]
> We should maybe explicitly document that it's not allowed to pass that FD to fdopendir.

Well if that's enough then we can also go the other direction without reopening. It's the question whether we want to tell users "don't do that" or rely on exclusive fd ownership to prevent it.

**Comment by RalfJung** [comment]
> Well if that's enough then we can also go the other direction without reopening

If we provide a way to call `fdopendir` then obviously that's not enough.

But yeah it is a function of whether we consider `fdopendir` to have the unsafe precondition of "underlying file description must be exclusively owned" or whether we consider it sound to do `fdopendir` on any `OwnedFd`.

But, anyway, for my needs a `dirfd`+`open` implementation would suffice. The point is that it's easy to implement `ReadDir::dir` in a way that is race-free on Unix and at least *works* on Windows, whereas achieving the same for `Dir::read_dir` is a lot harder.

So, that's the main remaining part of the question to me:
> And would it be okay to land an unstable first version that's only race-free on Unixes with dirfd? Of course the Windows version should be made race-free before stabilization.
